models/parser.py

```diff
@@ -0,0 +1,232 @@
+"""
+This file implements all classes needed to parse boolean expression.
+Here is the definition of our parser:
+    + Our alphabet is composed with:
+        - [a-zA-Z] letters
+        - &, |, , !, ( and ) symbols
+        - white space
+
+    + Our language is composed by 3 types of words:
+        - '!', '&&' and '||' operator
+        - '(' and ')' parenthesis
+        - [a-zA-Z]+ words
+
+    + Our grammar is defined here:
+        S::= Expression
+        Expression::= SubExpression
+                      | SubExpression '||' Expression  Operator: OR
+                      | SubExpression '&&' Expression  Operator: AND
+        SubExpression::= '('Expression')'
+                         | '!'SubExpression            Operator: NOT
+                         | RoleName
+        RoleName::= [a-zA-Z]+
```

Comment:
I don't really understand the difference between Expression and SubExpression here. The SubExpression is just the left term of the whole expression ? Maybe this should be detailed a little bit :-)

Reply:
The main difference is that Expression helps us to define AND and OR operators while the SubExpression defines NOT operator and parenthesis. This is only defined by our grammar.
I know this answer is not full of details but I don't really know how to explain better :p

Reply:
Ok I understood reading the code, for me it is mostly a way to separate the unary nodes (always a subExp) from the binary nodes (always an exp) to build the tree.